src/common/hist_util.cc

```diff
@@ -284,13 +300,14 @@ void ColsWiseBuildHistKernel(const std::vector<GradientPair> &gpair,
   };
 
   const size_t n_features = gmat.cut.Ptrs().size() - 1;
-  const size_t n_columns = n_features;
+  const size_t n_columns = kColumnSampling ? fids.size() : n_features;
```

Comment:
Is `fids` empty if there's no sampling?

Reply:
Fids is empty is case of condition [here](https://github.com/dmlc/xgboost/pull/8319/files#diff-8217fa4498a3341ec017b9661c7cc75123cc88fb058c865db9a7f71558a01cb0R91) is false.